services/rpcfilters/latest_block_changed_event.go

```diff
@@ -0,0 +1,117 @@
+package rpcfilters
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// latestBlockChangedEvent represents an event that one can subscribe to
+type latestBlockChangedEvent struct {
+	sx                  map[int]chan common.Hash
+	sxMu                sync.Mutex
+	previousLatestBlock blockInfo
+	latestBlockMu       sync.Mutex
+	provider            latestBlockProvider
+	quit                chan struct{}
+	tickerPeriod        time.Duration
+}
+
+func (e *latestBlockChangedEvent) Start() error {
+	if e.quit != nil {
+		return errors.New("latest block changed event is already started")
+	}
+
+	e.quit = make(chan struct{})
+
+	go func() {
+		ticker := time.NewTicker(e.tickerPeriod)
+		for {
+			select {
+			case <-ticker.C:
+				if e.numberOfSubscriptions() == 0 {
+					continue
+				}
+				latestBlock, err := e.provider.GetLatestBlock()
+				if err != nil {
+					log.Error("error while receiving latest block", "error", err)
+					continue
+				}
+
+				e.processLatestBlock(latestBlock)
+			case <-e.quit:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (e *latestBlockChangedEvent) numberOfSubscriptions() int {
+	e.sxMu.Lock()
+	defer e.sxMu.Unlock()
+	return len(e.sx)
+}
+
+func (e *latestBlockChangedEvent) processLatestBlock(latestBlock blockInfo) {
+	e.latestBlockMu.Lock()
+	defer e.latestBlockMu.Unlock()
+
+	// if we received the same or an older block than we already have, ignore it.
+	if latestBlock.Number().Cmp(e.previousLatestBlock.Number()) <= 0 {
```

Comment:
it might be a bug. how original implementation behaves if chain reorgs? it is definitely possible that multiple blocks will be mined on the same chain height

Reply:
that is caused by infura's load balancing.

Reply:
no, chain reorganization happens because ethereum doesn't provide instant finality for a block

https://blog.ethereum.org/2015/08/08/chain-reorganisation-depth-expectations/

Reply:
I'm talking about that specific case. That could be easily seen if you request the latest block by RPC.
If the chain is A1, A2, A3.
You will (most probably) receive something like: A1, A2, A1, A2, A3, A2, A3, A3, A3, A3...
That is not the overwrites, because it is the same blocks.

Reply:
but you will discard any overwrite with that logic, so for example you tracked that the last block is:

number = 101 : hash = aaa

but then chain reorg happens and :

number = 101 : hash = bbb -> number = 102 : hash = zzz

and the last one is actually becomes canonical chain because it turned out to be longer, but you will never notify user about block with hash bbb